File: memory/snapshot_engine.py
import json
import hashlib
from typing import Dict, Any, Optional
from core.logger import logger

class SnapshotEngine:
    """
    Immutable, content-addressable memory state snapshots.
    """
    def __init__(self, storage_adapter):
        self.storage = storage_adapter
        
    def save_snapshot(self, state: Dict[str, Any]) -> str:
        state_str = json.dumps(state, sort_keys=True)
        h = hashlib.sha256(state_str.encode('utf-8')).hexdigest()
        
        # Objects addressed by hash
        self.storage.save(f"snapshot_{h}", state)
        logger.info("Snapshot saved to storage", snapshot_hash=h)
        return h
        
    def get_snapshot(self, snapshot_hash: str) -> Optional[Dict[str, Any]]:
        logger.debug("Loading snapshot from storage", snapshot_hash=snapshot_hash)
        return self.storage.load(f"snapshot_{snapshot_hash}")

File: sdk/typescript/src/memory/snapshot_engine.ts
import { createHash } from 'crypto';
import { StorageAdapter } from '../storage/adapters';
import { logger } from '../core/logger';

export class SnapshotEngine {
    private storage: StorageAdapter;

    constructor(storage: StorageAdapter) {
        this.storage = storage;
    }

    private hashDict(data: Record<string, any>): string {
        const sorted = JSON.stringify(data, Object.keys(data).sort());
        return createHash('sha256').update(sorted).digest('hex');
    }

    async saveSnapshot(data: Record<string, any>, parentHash: string | null = null): Promise<string> {
        const snapshotHash = this.hashDict(data);
        const payload = {
            snapshot_hash: snapshotHash,
            parent_hash: parentHash,
            data: data,
            timestamp: Date.now() / 1000.0
        };
        await this.storage.save(`snap_${snapshotHash}`, payload);
        logger.info('Snapshot saved to storage', { snapshot_hash: snapshotHash });
        return snapshotHash;
    }

    async loadSnapshot(snapshotHash: string): Promise<Record<string, any>> {
        const snap = await this.storage.load(`snap_${snapshotHash}`);
        if (!snap) throw new Error(`Snapshot not found: ${snapshotHash}`);
        return snap.data;
    }
}
